tests/test_factory.py

```diff
@@ -1060,3 +1060,130 @@ def test_poetry_build_system_dependencies(
     poetry = Factory().create_poetry(temporary_directory)
 
     assert set(poetry.build_system_dependencies) == expected
+
+
+def test_create_poetry_with_nested_dependency_groups(temporary_directory: Path) -> None:
+    pyproject_toml = temporary_directory / "pyproject.toml"
+    content = """\
+[project]
+name = "my-package"
+version = "1.2.3"
+
+[tool.poetry.group.testing.dependencies]
+pytest = "*"
+pytest-cov ="*"
+
+[tool.poetry.group.dev]
+include-groups = [
+    "testing",
+]
+[tool.poetry.group.dev.dependencies]
+black = "*"
+"""
+    pyproject_toml.write_text(content)
+    poetry = Factory().create_poetry(temporary_directory)
+
+    assert len(poetry.package.all_requires) == 5
+    assert [
+        (dep.name, ",".join(dep.groups)) for dep in poetry.package.all_requires
+    ] == [
+        ("pytest", "testing"),
+        ("pytest-cov", "testing"),
+        ("black", "dev"),
+        ("pytest", "dev"),
+        ("pytest-cov", "dev"),
+    ]
+
+
+def test_create_poetry_with_invalid_nested_dependency_groups(
```

Comment:
We should test three variants of this:

- self-reference
- direct cycle (that is what is tested now)
- indirect cycle (with three groups)

(and as mentioned in the other comment improve the error message)

Reply:
Test case added.